# **INTRODUCTION**
In India, increased urbanisation, industrial activity, and increasing car emissions all contribute to dangerous air quality levels, making air pollution a serious public health concern.Understanding how various pollutants affect overall air quality in Indian cities is crucial since high concentrations of toxic gases and particulate matter are associated with major health problems. An air-quality dataset collected from different monitoring sites in several Indian cities is analysed in this study. The goal is to determine which pollutants are most responsible for poor air quality and to offer information that improves with efficient environmental monitoring. This project will be employing a comprehensive data-analysis methodology that includes data cleaning, exploratory data analysis (EDA), and a basic predictive model in order to accomplish this. GitHub is used for version control, and visualisations and an interactive Streamlit application are used to convey the results.



# **Importing the required libraries**

In [5]:
import pandas as pd
import numpy as np
import os
import matplotlib as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go

# **Mounting the drive**

In [6]:
from google.colab import drive

In [7]:
drive.mount('/content/drive')

Mounted at /content/drive


In [8]:
drive_path = '/content/drive/MyDrive/GOOGLE COLAB PROJECTS/PROGRAMMING FOR DATA ANALYSIS ASSIGNMENT/Assessment Data-20251105'

# **Combining all the csv files in the assessment data folder in the drive path**

In order to analyse the pollutants consistently across all locations and time periods, the dataset, which is divided into several CSV files, must be integrated into a single dataframe. Complete and accurate data cleaning, exploration, and modelling are made possible by merging the files.




In [9]:
import os
import pandas as pd

dataframes = []

for filename in os.listdir(drive_path):
    if filename.endswith('.csv'):
        file_path = os.path.join(drive_path, filename)
        df = pd.read_csv(file_path)
        dataframes.append(df)

full_df = pd.concat(dataframes, ignore_index=True)

# **Investigating the Combined Dataset**

I will now look at the first and last rows, verify the data types, then use head(), tail(), info(), and describe() to study summary statistics in order to comprehend the structure and contents of the combined dataset.

In [10]:
full_df.head()

City        Date  PM2.5  PM10  NO  NO2  NOx  NH3    CO  SO2    O3  \
0  Talcher  20/12/2017    NaN   NaN NaN  NaN  NaN  NaN  3.60  NaN  0.02   
1  Talcher  21/12/2017    NaN   NaN NaN  NaN  NaN  NaN  3.41  NaN  0.03   
2  Talcher  22/12/2017    NaN   NaN NaN  NaN  NaN  NaN  5.15  NaN  0.02   
3  Talcher  23/12/2017    NaN   NaN NaN  NaN  NaN  NaN   NaN  NaN   NaN   
4  Talcher  24/12/2017    NaN   NaN NaN  NaN  NaN  NaN   NaN  NaN   NaN   

   Benzene  Toluene  Xylene  AQI AQI_Bucket  
0      NaN      NaN     NaN  NaN        NaN  
1      NaN      NaN     NaN  NaN        NaN  
2      NaN      NaN     NaN  NaN        NaN  
3      NaN      NaN     NaN  NaN        NaN  
4      NaN      NaN     NaN  NaN        NaN

In [11]:
full_df.tail()

City        Date  PM2.5   PM10     NO   NO2    NOx    NH3    CO  \
29526  Mumbai  27/06/2020  18.04  63.72  28.29  8.58  36.07  26.69  0.43   
29527  Mumbai  28/06/2020  12.33  44.84  27.15  8.06  34.78  26.50  0.38   
29528  Mumbai  29/06/2020   7.88  27.51  26.58  7.94  34.42  23.94  0.38   
29529  Mumbai  30/06/2020   4.66  13.87  25.25  8.02  33.25  19.89  0.36   
29530  Mumbai  01/07/2020   3.84  11.28  27.64  5.57  33.21  23.60  0.37   

         SO2    O3  Benzene  Toluene  Xylene   AQI    AQI_Bucket  
29526  10.36  5.68     0.82      NaN     NaN  64.0  Satisfactory  
29527  10.58  5.76     0.85      NaN     NaN  64.0  Satisfactory  
29528  11.11  5.84     0.47      NaN     NaN  54.0  Satisfactory  
29529  13.40  6.94     0.73      NaN     NaN  54.0  Satisfactory  
29530  13.18  6.68     0.78      NaN     NaN  53.0  Satisfactory

In [12]:
full_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29531 entries, 0 to 29530
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   City        29531 non-null  object 
 1   Date        29531 non-null  object 
 2   PM2.5       24933 non-null  float64
 3   PM10        18391 non-null  float64
 4   NO          25949 non-null  float64
 5   NO2         25946 non-null  float64
 6   NOx         25346 non-null  float64
 7   NH3         19203 non-null  float64
 8   CO          27472 non-null  float64
 9   SO2         25677 non-null  float64
 10  O3          25509 non-null  float64
 11  Benzene     23908 non-null  float64
 12  Toluene     21490 non-null  float64
 13  Xylene      11422 non-null  float64
 14  AQI         24850 non-null  float64
 15  AQI_Bucket  24850 non-null  object 
dtypes: float64(13), object(3)
memory usage: 3.6+ MB


In [13]:
full_df.describe()

PM2.5          PM10            NO           NO2           NOx  \
count  24933.000000  18391.000000  25949.000000  25946.000000  25346.000000   
mean      67.450578    118.127103     17.574730     28.560659     32.309123   
std       64.661449     90.605110     22.785846     24.474746     31.646011   
min        0.040000      0.010000      0.020000      0.010000      0.000000   
25%       28.820000     56.255000      5.630000     11.750000     12.820000   
50%       48.570000     95.680000      9.890000     21.690000     23.520000   
75%       80.590000    149.745000     19.950000     37.620000     40.127500   
max      949.990000   1000.000000    390.680000    362.210000    467.630000   

                NH3            CO           SO2            O3       Benzene  \
count  19203.000000  27472.000000  25677.000000  25509.000000  23908.000000   
mean      23.483476      2.248598     14.531977     34.491430      3.280840   
std       25.684275      6.962884     18.133775     21.694928     15.811136   
min        0.010000      0.000000      0.010000      0.010000      0.000000   
25%        8.580000      0.510000      5.670000     18.860000      0.120000   
50%       15.850000      0.890000      9.160000     30.840000      1.070000   
75%       30.020000      1.450000     15.220000     45.570000      3.080000   
max      352.890000    175.810000    193.860000    257.730000    455.030000   

            Toluene        Xylene           AQI  
count  21490.000000  11422.000000  24850.000000  
mean       8.700972      3.070128    166.463581  
std       19.969164      6.323247    140.696585  
min        0.000000      0.000000     13.000000  
25%        0.600000      0.140000     81.000000  
50%        2.970000      0.980000    118.000000  
75%        9.150000      3.350000    208.000000  
max      454.850000    170.370000   2049.000000

# **Data understanding/Data Overview**

In this dataset here are 29,531 rows and 16 columns of the combined dataset. It represents the hourly air-quality measurements taken in a number of Indian cities. There are thirteen numerical pollutant variables (PM2.5, PM10, NO, NO2, NOx, NH3, CO, SO2, O3, Benzene, Toluene, Xylene, and AQI) and three categorical features (City, Date, and AQI_Bucket) are included in this combined dataset. Additionally, a number of pollutant columns in the info() output have missing values, which will have to be rectified during preprocessing. This dataset supports the project's objective of identifying the pollutants that contribute to poor air quality by offering a comprehensive perspective of pollutant concentrations and air-quality conditions.

# **Data preprocessing:**

**Handling missing data**

In [14]:
def missing_values_table(df):
    mis_val = df.isnull().sum()
    mis_val_percent = 100 * mis_val / len(df)

    mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
    mis_val_table = mis_val_table.rename(columns={
        0: 'Missing Values',
        1: '% of Total Values'
    })

    mis_val_table = mis_val_table.sort_values('% of Total Values', ascending=False)

    return mis_val_table

missing_values = missing_values_table(full_df)
display(missing_values.style.background_gradient(cmap='Greens'))


**Missing value summary:**

I started by looking at the quantity and percentage of missing values in each column to determine the dataset's quality. This makes it easier to determine which features could require cleaning or removal and which pollutant measures are lacking. The findings demonstrate that while some pollutants, like CO and NO, have comparatively few missing values, others, like Xylene, PM10, and NH3, have substantial amounts of missing data. On the other hand, there are no missing values in the City and Date columns, which indicates that the time and location data are complete, however it is not necessary because it is not essential to my research question.

**Identifying Pollutant Columns for Analysis:**

Since my research question aims to identify which pollutant most strongly contributes to poor air quality, I began by identifying all numerical pollutant variables in the dataset. These selected columns represent measured concentrations of various air contaminants across Indian cities and will be used throughout the cleaning, exploratory analysis, and modelling stages.

In [15]:
full_df.columns


Index(['City', 'Date', 'PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2',
       'O3', 'Benzene', 'Toluene', 'Xylene', 'AQI', 'AQI_Bucket'],
      dtype='object')

In [16]:
pollutant_cols = [
    'PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3',
    'CO', 'SO2', 'O3', 'Benzene', 'Toluene',
    'Xylene', 'AQI'
]


# **Removing duplicate entries**

I started by looking for duplicate entries to make sure the dataset was clean and credible. By increasing pollution levels deliberately, duplicate pollutant measurements might distort the results of my research. Accuracy is crucial since my study topic seeks to determine which pollutant is most responsible for poor air quality. I first checked how many duplicates existed in the dataset which was 0 which indicates that no rows were repeated. I used drop_duplicates() as a precaution to make sure the dataset stayed clean even if there were no duplicates. After cleaning, the final check verified that there were no duplicates left. This guarantees that every pollutant measurement makes a fair and accurate contribution to the research.

**Checking for duplicates:**


In [17]:
# Step 1: Check how many duplicate rows exist
print(f"There are {full_df.duplicated().sum()} duplicate rows in the dataset.")

There are 0 duplicate rows in the dataset.


**Removing duplicates:**

In [18]:
# Step 2: Remove duplicate rows
full_df = full_df.drop_duplicates()

print("Duplicates removed.")

Duplicates removed.


**Final check of removal:**

In [19]:
print(f"Duplicates left after cleaning: {full_df.duplicated().sum()}")

Duplicates left after cleaning: 0


# **Feature engineering**
In order to make the analysis more understandable and informative, feature engineering entails preparing and enhancing the dataset by adding or changing variables. The feature engineering procedures are straightforward and directly applicable because my research issue focusses on determining which pollutant is most responsible for bad air quality.

The main tasks include:

1. Making the date column in the corerct datatime format

Even though my study is concentrated on pollutants, this enables me to extract relevant time-based information later (e.g., month, year).

In [32]:
full_df['Date'] = pd.to_datetime(full_df['Date'], errors='coerce')

2. Extracting the month and year

Adding a month column aids in understanding pollutant fluctuation over time, even though seasonality is not my primary study objective.

In [33]:
full_df['Month'] = full_df['Date'].dt.month

3. Choosing the relevant pollutants

I previously made a list in my notebook named pollutant_cols that included every numerical pollutant variable significant to my research issue. I will now utilise this previously created list to extract only the pollutant-related features from the dataset because my goal is to determine which pollutant contributes most to bad air quality. By concentrating on the behaviour of pollutants, this stage produces a cleaner subset of the data. The remaining cleaning, modelling, and visualisation procedures will be performed using this subset (pollutant_df). However, I made an separate version of the pollutant dataset using.copy(). This just keeps the original dataset (full_df) from being accidentally altered when cleaning or altering pollutant_df.

In [34]:
pollutant_df = full_df[pollutant_cols].copy()